server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java

```diff
@@ -1040,8 +1041,88 @@ public void testDoNotRebalanceToTheNodeThatNoLongerExists() {
         assertThat(allocation.routingNodes().node("data-node-2").getByShardId(shardId), notNullValue());
     }
 
+    public void testRebalanceDoesNotCauseHotSpots() {
```

Comment:
This test looks quite general but is in fact very specific: it needs the initial allocation of shards to be such that all the shards on each initial node must move to the same target node. It's then saying that we round-robin through the nodes rather than focussing on a single node's shards. I think I'd prefer it to be more explicit about this - rather than saying that the cluster never sees a hotspot, I think we should count the number of times a shard moves off each node and assert that these numbers don't get too far apart.